homeassistant/components/withings/api.py

```diff
@@ -0,0 +1,157 @@
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.config_entry_oauth2_flow import (
+    AbstractOAuth2Implementation,
+    OAuth2Session,
+)
+
+from .const import LOG_NAMESPACE
+
+_LOGGER = logging.getLogger(LOG_NAMESPACE)
+_RETRY_COEFFICIENT = 0.5
+
+
+class ConfigEntryWithingsApi(AbstractWithingsApi):
+    """Withing API that uses HA resources."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        config_entry: ConfigEntry,
+        implementation: AbstractOAuth2Implementation,
+    ) -> None:
+        """Initialize object."""
+        self._hass = hass
+        self.config_entry = config_entry
+        self._implementation = implementation
+        self.session = OAuth2Session(hass, config_entry, implementation)
+
+    def _request(
+        self, path: str, params: dict[str, Any], method: str = "GET"
+    ) -> dict[str, Any]:
+        """Perform an async request."""
+        asyncio.run_coroutine_threadsafe(
+            self.session.async_ensure_token_valid(), self._hass.loop
+        ).result()
+
+        access_token = self.config_entry.data["token"]["access_token"]
+        response = requests.request(
+            method,
+            f"{self.URL}/{path}",
+            params=params,
+            headers={"Authorization": f"Bearer {access_token}"},
+            timeout=10,
+        )
+        return response.json()
+
+    async def _do_retry(self, func, attempts=3) -> Any:
+        """Retry a function call.
+
+        Withings' API occasionally and incorrectly throws errors.
+        Retrying the call tends to work.
+        """
+        exception = None
+        for attempt in range(1, attempts + 1):
+            _LOGGER.debug("Attempt %s of %s", attempt, attempts)
+            try:
+                return await func()
+            except Exception as exception1:  # pylint: disable=broad-except
+                _LOGGER.debug(
+                    "Failed attempt %s of %s (%s)", attempt, attempts, exception1
+                )
+                # Make each backoff pause a little bit longer
+                await asyncio.sleep(_RETRY_COEFFICIENT * attempt)
+                exception = exception1
+                continue
+
+        if exception:
+            raise exception
+
+    async def async_measure_get_meas(
+        self,
+        meastype: MeasureType | None = None,
+        category: MeasureGetMeasGroupCategory | None = None,
+        startdate: DateType | None = arrow.utcnow(),
+        enddate: DateType | None = arrow.utcnow(),
+        offset: int | None = None,
+        lastupdate: DateType | None = arrow.utcnow(),
+    ) -> MeasureGetMeasResponse:
+        """Get measurements."""
+
+        def call_super() -> MeasureGetMeasResponse:
+            return self.measure_get_meas(
+                meastype, category, startdate, enddate, offset, lastupdate
+            )
+
+        return await self._do_retry(await self._hass.async_add_executor_job(call_super))
```

Comment:
Instead of defining the call_super method, you could have just directly called the method and added the arguments like:
`async_add_executor_job(self.measure_get_meas, meastype, category, startdate, enddate, offset, lastupdate)`

But not sure how that looks like after black formatting of the long line.

Reply:
I mean, it doesn't make it more readable, but I am planning to remove the whole file so I don't really mind

Reply:
I personally like it better now, but it does not really matter.